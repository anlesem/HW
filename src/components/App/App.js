import { useState, useEffect } from 'react';
import { nanoid } from 'nanoid';

import style from './App.module.scss';

import { Header } from '../Header/Header';
// import { Message } from '../Message/Message';
import { MessageList } from '../MessageList/MessageList';
import { Form } from '../Form/Form';
import { Footer } from '../Footer/Footer';


function App() {
	const [messages, setMessages] = useState([]);
	// const robotRef = useRef(null);

	// text, event для разнообразия, но лучше не использовать
	const sendMassage = (text, event) => {
		if (event.target.elements.author.value || event.target.elements.text.value) {
			setMessages([...messages, {
				id: nanoid(),
				author: text,
				text: event.target.elements.text.value,
			}]);
			// Так делать плохо. Необходимо очищать через setState
			event.target.elements.text.value = '';
		}
		// console.log(messages);	// вернёт старое значение
	}

	useEffect(() => {
		// console.log(messages);	// вернёт новое значение
	}, [messages]);


	return (
		<div className={style.root}>
			<Header />
			<main className="container">
				{/* <Message content={'Все задания выполнены'} /> */}
				<Form addMessage={sendMassage} />
				<MessageList message={messages} />
			</main>
			<Footer />
		</div>
	);
}

export default App;
