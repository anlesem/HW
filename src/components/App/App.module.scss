.root {
	min-height: 100%;
	display: grid;
	grid-template-rows: 100px 1fr 100px;
}

.main {
	position: relative;
	width: 100%;

	display: flex;
	justify-content: space-around;
}