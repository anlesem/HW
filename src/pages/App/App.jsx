import React, { useEffect } from 'react';
import { BrowserRouter, Route, Routes } from 'react-router-dom';
import { useDispatch } from 'react-redux';

import style from './App.module.scss';

import { Layout } from '../Layout/Layout';
import { Profile } from '../Profile/Profile';
import { Data } from '../Data/Data';
import { NotFound } from '../NotFound/NotFound';
import { ProgressBar } from '../../components/ProgressBar/ProgressBar';
import { Sign } from '../Sign/Sign';
import { auth } from '../../services/firebase';
import { changeName, toggleLogin } from '../../store/profile/actions';

// const Chats = React.lazy(() => import('../Chats/Chats'));
const Chats = React.lazy(() =>
  import('../Chats/Chats').then((module) => ({ default: module.Chats }))
);

export default function App() {
  const dispatch = useDispatch();

  useEffect(() => {
    auth.onAuthStateChanged((user) => {
      if (user) {
        dispatch(toggleLogin(true));
        dispatch(changeName(user.email));
      } else {
        dispatch(toggleLogin(false));
        dispatch(changeName('user'));
      }
    });
  }, []);

  return (
    <div className={style.root}>
      <BrowserRouter>
        <Routes>
          <Route path="/" element={<Layout />}>
            <Route index element={<Profile />} />
            <Route path="sign" element={<Sign />} />
            <Route path="sign/:reg" element={<Sign />} />
            <Route
              path="chats"
              element={
                <React.Suspense fallback={<ProgressBar />}>
                  <Chats />
                </React.Suspense>
              }
            />
            <Route
              path="chats/:chatId"
              element={
                <React.Suspense fallback={<ProgressBar />}>
                  <Chats />
                </React.Suspense>
              }
            />
            <Route path="data" element={<Data />} />
            <Route path="*" element={<NotFound />} />
          </Route>
        </Routes>
      </BrowserRouter>
    </div>
  );
}
