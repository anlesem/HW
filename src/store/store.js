import { createStore, combineReducers, compose, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';
import { persistStore, persistReducer } from 'redux-persist';
import storage from 'redux-persist/lib/storage';

import { profileReducer } from './profile/reducer';
import { chatsReducer } from './chats/reducer';
import { messagesReducer } from './messages/reducer';
import { dataReducer } from './data/reducer';

const persistConfig = {
  key: 'root',
  storage,
  blacklist: ['data']
};

const rootReducers = combineReducers({
  profile: profileReducer,
  chats: chatsReducer,
  messages: messagesReducer,
  data: dataReducer
});

const persistedReducers = persistReducer(persistConfig, rootReducers);

export const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

export const store = createStore(persistedReducers, composeEnhancers(applyMiddleware(thunk)));

export const persistor = persistStore(store);
