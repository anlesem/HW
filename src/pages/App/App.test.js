import { render, screen, fireEvent, cleanup } from '@testing-library/react';

import App from './App';

import { Provider } from 'react-redux';
import { store } from '../../store/store';

describe('App базис', () => {
  it('Компонент существует', () => {
    expect(App).toBeInstanceOf(Function);
  });

  it('Снимок состояния', () => {
    const { asFragment } = render(
      <Provider store={store}>
        <App />
      </Provider>
    );
    expect(asFragment(<App />)).toMatchSnapshot();
  });
});

describe('App переходы и авторизация', () => {
  beforeEach(() => {
    render(
      <Provider store={store}>
        <App />
      </Provider>
    );
  });

  afterEach(cleanup);

  it('Состояние при загрузке', () => {
    expect(screen.getByText(/чаты/i)).toBeTruthy();
    expect(screen.getByText(/профиль/i)).toBeTruthy();
    expect(screen.getByText(/данные/i)).toBeTruthy();
    expect(screen.getAllByRole('heading')).toBeTruthy();
  });

  it('Переходы между страницами', async () => {
    const navLink1 = screen.getByTestId('NavLink-1');
    const navLink2 = screen.getByTestId('NavLink-2');
    const navLink3 = screen.getByTestId('NavLink-3');

    expect(screen.getByText(/user/i)).toBeTruthy();

    fireEvent.click(navLink2);

    expect(screen.getByText(/авторизация/i)).toBeTruthy();

    fireEvent.click(navLink3);

    expect(screen.getByTestId('progress')).toBeTruthy();

    fireEvent.click(navLink1);

    expect(screen.getByText(/user/i)).toBeTruthy();

    fireEvent.click(screen.getByTestId('profile-link-enter'));

    expect(screen.getByText(/авторизация/i)).toBeTruthy();
    expect(screen.getByText(/вход/i)).toBeTruthy();

    fireEvent.click(navLink1);
    fireEvent.click(screen.getByTestId('profile-link-reg'));

    expect(screen.getByText(/авторизация/i)).toBeTruthy();
    expect(screen.getByText(/регистрация/i)).toBeTruthy();
  });

  it('Авторизация. Поля', () => {
    expect(screen.getByText(/авторизация/i)).toBeTruthy();
    expect(screen.getByText(/регистрация/i)).toBeTruthy();

    const login = screen.getByTestId('sign-email');
    const button = screen.getByTestId('sign-submit');

    fireEvent.change(login, { target: { value: 'Вася' } });
    expect(login.value).toBe('Вася');

    fireEvent.click(button);

    expect(screen.getByText(/авторизация/i)).toBeTruthy();
    expect(screen.getByText(/регистрация/i)).toBeTruthy();
  });

  it('Регистрация', async () => {
    expect(screen.getByText(/авторизация/i)).toBeTruthy();
    expect(screen.getByText(/регистрация/i)).toBeTruthy();

    const login = screen.getByTestId('sign-email');
    const password = screen.getByTestId('sign-password');
    const button = screen.getByTestId('sign-submit');

    fireEvent.change(login, { target: { value: 'noLogin@test.test' } });
    fireEvent.change(password, { target: { value: '1234' } });
    fireEvent.click(button);

    await new Promise((res) => setTimeout(res, 1000));

    expect(screen.getByText(/Password should be at least 6 characters/)).toBeTruthy();

    fireEvent.change(login, { target: { value: 'login@test.test' } });
    fireEvent.change(password, { target: { value: '123456' } });
    fireEvent.click(button);

    await new Promise((res) => setTimeout(res, 1000));

    expect(screen.getByText(/auth\/email-already-in-use/)).toBeTruthy();
  });

  it('Вход', async () => {
    fireEvent.click(screen.getByTestId('NavLink-1'));
    fireEvent.click(screen.getByTestId('profile-link-enter'));

    expect(screen.getByText(/вход/i)).toBeTruthy();

    const login = screen.getByTestId('sign-email');
    const password = screen.getByTestId('sign-password');
    const button = screen.getByTestId('sign-submit');

    fireEvent.change(login, { target: { value: 'noLogin@test.test' } });
    fireEvent.change(password, { target: { value: '123456' } });
    fireEvent.click(button);

    await new Promise((res) => setTimeout(res, 1000));

    expect(screen.getByText(/auth\/user-not-found/)).toBeTruthy();

    fireEvent.change(login, { target: { value: 'login@test.test' } });
    fireEvent.change(password, { target: { value: '12345678' } });
    fireEvent.click(button);

    await new Promise((res) => setTimeout(res, 1000));

    expect(screen.getByText(/auth\/wrong-password/)).toBeTruthy();

    fireEvent.change(login, { target: { value: 'login@test.test' } });
    fireEvent.change(password, { target: { value: '123456' } });
    fireEvent.click(button);

    await new Promise((res) => setTimeout(res, 1000));

    expect(screen.getByText('login@test.test')).toBeTruthy();
  });

  it('Выход', async () => {
    fireEvent.click(screen.getByTestId('profile-button-exit'));

    await new Promise((res) => setTimeout(res, 1000));

    expect(screen.getByText('email')).toBeTruthy();
  });

  it('Авторизация в Чатах', async () => {
    fireEvent.click(screen.getByTestId('NavLink-2'));

    expect(screen.getByText(/вход/i)).toBeTruthy();

    const login = screen.getByTestId('sign-email');
    const password = screen.getByTestId('sign-password');
    const button = screen.getByTestId('sign-submit');

    fireEvent.change(login, { target: { value: 'login@test.test' } });
    fireEvent.change(password, { target: { value: '123456' } });
    fireEvent.click(button);

    await new Promise((res) => setTimeout(res, 2000));

    expect(screen.getByText('Сообщения:')).toBeTruthy();
  });
});
