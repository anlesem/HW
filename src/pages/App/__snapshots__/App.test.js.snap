// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Снимок состояния 1`] = `
<DocumentFragment>
  <div
    class="root"
  >
    <header
      class="header"
    >
      <div
        class="wrap container"
      >
        <h1>
          Урок 5
        </h1>
        <ul
          class="navbar"
        >
          <li>
            <a
              aria-current="page"
              class="active"
              data-testid="NavLink-1"
              href="/"
            >
              Профиль
            </a>
          </li>
          <li>
            <a
              class=""
              data-testid="NavLink-2"
              href="/chats"
            >
              Чаты
            </a>
          </li>
        </ul>
      </div>
    </header>
    <div
      class="main"
    >
      <h1
        class="head"
      >
        user
      </h1>
      <label
        class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-j204z7-MuiFormControlLabel-root"
        data-testid="profile-box"
      >
        <span
          class="MuiCheckbox-root MuiCheckbox-colorPrimary MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-root css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
        >
          <input
            class="PrivateSwitchBase-input css-1m9pwf3"
            data-indeterminate="false"
            data-testid="profile-check"
            type="checkbox"
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="CheckBoxOutlineBlankIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </span>
        <span
          class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-ahj2mt-MuiTypography-root"
        >
          Авторизоваться
        </span>
      </label>
    </div>
    <footer
      class="footer"
    >
      <div
        class="wrap container"
      >
        <p
          class="date"
          data-testid="footer-paragraph"
        >
          © 2022
        </p>
      </div>
    </footer>
  </div>
</DocumentFragment>
`;
